curp/doc/steps.md

```diff
@@ -0,0 +1,121 @@
+The document describes the CURP steps in detail.
+
+Definition:
+1. The request tuple consists of request command and following parameters.
+2. The response tuple consists of return code and following states.
+
+# Normal Operation
+## A. Client Send Propose
+
+1. Client send request <Propose, Cmd, RpcId, TermId, WitnessListVersion> to the master and all witnesses.
+
+Note:
+- `TermId` works the same way as in the Raft, which tells the leader version.
+- `RpcId` is the universal id used for the command request. And the resent request reuse the RpcId to avoid repeated execution.
+- `WitnessListVersion` is used in witness change during membership change. When witness set changes, its value increases 1.
+- As `TermId` is used in all request and response, we don't put it in the following tuples.
+
+## B. Master Handle Client Propose
+
+Request: <Propose, Cmd, RpcId, WitnessListVersion>
+
+1. Check `TermId`
+    - if `TermId` doesn't match, reply tuple <WrongTermId> back to the client. 
+    - otherwise continue.
+2. Check `WitnessListVersion`
+    - if `WitnessListVersion` doesn't match, reply tuple <WrongWitnessListVersion, WitnessListVersion>
```

Comment:
also reply with the current witness list?

Reply:
Witness list comes from the configuration server, so we don't have to reply the witness list itself.